docs: add verification documentation (#3407)
akuity/kargo: docs/docs/60-new-docs/50-user-guide/20-how-to-guides/16-verification.md

---
description: Learn how to verify a Stage after Promotion
sidebar_label: Verifying Freight in a Stage
---

# Verifying a Stage after Promotion

> After successfully promoting a Freight into a Stage, it is often desired to verify the update was successful. Verification is an optional, user-defined process that qualifies Freight in a Stage. Common examples of verification include:

Reviewer: ```suggestion
After successfully promoting `Freight` to a `Stage`, it is often desired to verify the update was successful. Verification is an optional, user-defined process that qualifies `Freight` in a `Stage`. Common examples of verification include:

```

* Executing containerized integration test scripts
* Querying monitoring tools (e.g. DataDog, Prometheus) and measuring application KPIs
* Performing HTTP requests to internal services or systems and interpreting response as success/fail
* Running load generation tests

> When a Stage with verification acts as an upstream to other Stages, Freight must pass verification in that Stage before it is eligible to be promoted to the downstream Stages in the pipeline.

Reviewer: ```suggestion
When a `Stage` with verification is upstream from other `Stages`, `Freight` must pass verification there before becoming eligible to be promoted downstream.
```

## Configuring Verification

> A Stage can be configured with verification by referencing one or more `AnalysisTemplates` in its `spec.verification` field. In the following example, the `dev` `Stage` references an `AnalysisTemplate` named `integration-test`, which executes after any successful Promotion.

Reviewer: ```suggestion
A `Stage` can be configured with verification by referencing one or more `AnalysisTemplates` in its `spec.verification` field. In the following example, the `dev` `Stage` references an `AnalysisTemplate` named `integration-test`, which executes after any successful `Promotion`:
```

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: dev
  namespace: guestbook
spec:
  # ...
  verification:
    analysisTemplates:
    - name: integration-test
```

## Configuring AnalysisTemplates

> An [AnalysisTemplate](../60-reference-docs/50-analysis-templates.md) defines instructions how to perform verification. It might include running a Kubernetes Job, query monitoring tools, interpreting their result, performing HTTP requests, or any combination of these actions. AnalysisTemplates reside in the same `Project` Namespace of the `Stage` from which they are referenced. The following simple `AnalysisTemplate` example will create a Kubernetes `Job` that runs an alpine container that sleeps for 10 seconds.

Reviewer: ```suggestion
An [AnalysisTemplate](../60-reference-docs/50-analysis-templates.md) defines instructions for how to perform verification. It might include running a Kubernetes `Job`, querying monitoring tools, interpreting their result, performing HTTP requests, or any combination of these actions. `AnalysisTemplate`s reside in the same `Project` namespace as the `Stage` from which they are referenced. The following simple `AnalysisTemplate` example will create a Kubernetes `Job` that runs an alpine container that sleeps for 10 seconds:
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: AnalysisTemplate
metadata:
  name: integration-test
  namespace: guestbook
spec:
  metrics:
  - name: integration-test
    provider:
      job:
        spec:
          template:
            spec:
              containers:
              - name: sleep
                image: alpine:latest
                command: [sleep, "10"]
              restartPolicy: Never
          backoffLimit: 1
```

:::info

> For further documentation and examples of AnalysisTemplates, refer to the

Reviewer: ```suggestion
For further documentation and examples of `AnalysisTemplate`s, refer to the
```

[`AnalysisTemplate` reference](../60-reference-docs/50-analysis-templates.md).
:::

## How Verification Works

Upon successful promotion, a Stage enters the `Verifying` phase. Any AnalysisTemplates which were referenced in the Stage are invoked into an resource called an `AnalysisRun`, which represents a single invocation of the templates into a batch-like object. 

> When the `AnalysisRun` completes and is `Successful`, the verification is considered successful and Freight is marked as "verified" in that Stage. This makes it possible for the Freight to promoted to any immediate downstream Stages, either automatically (with auto-promotion) or manually (via UI, API, CLI). Freight which have not passed verification in a Stage are blocked from being able to continue to downstream Stages.

Reviewer: ```suggestion
When the `AnalysisRun` completes and is `Successful`, the verification is considered successful and the `Freight` is marked as "verified" in that `Stage`, which makes it available for promotion to any immediate downstream `Stage`s, either automatically (with auto-promotion) or manually (via UI, API, CLI). `Freight` which have not passed verification in a `Stage` are blocked continuing further downstream.
```

:::tip

> It is sometimes desired to skip the verification process entirely (e.g. a hotfix needs to be fast-tracked to production). Verification (and the entire Pipeline for that matter) can be bypassed by [manually approving](working-with-freight#manual-approvals) a Freight for a specified Stage.

Reviewer: ```suggestion
It is sometimes desired to skip the verification process entirely (e.g. a hotfix needs to be fast-tracked to production). Verification (and the entire Pipeline for that matter) can be bypassed by [manually approving](working-with-freight#manual-approvals) `Freight` for a specified `Stage`.
```

:::

> While a Stage is verifying (i.e. the `AnalysisRun` is not yet completed) no new Promotions will run until verification has completed (either successfully or unsuccessfully). Once verification has completed, the next queued Promotion will run.

Reviewer: ```suggestion
While a `Stage` is verifying (i.e. the `AnalysisRun` is not yet completed) no other `Promotion`s to that `Stage` will run until verification has completed (either successfully or unsuccessfully). Once verification has completed, the next queued `Promotion` will run.
```

## AnalysisRun

> An AnalysisRun is a resource representing a verification attempt for a combination of Freight and Stage. AnalysisRuns have a lifecycle (e.g. Running, Successful, Failed) and record the measurements taken during the run. AnalysisRuns appear in the Kargo UI under the Verifications tab of a stage:

Reviewer: ```suggestion
An `AnalysisRun` is a resource representing a verification attempt for a `Freight` + `Stage` pair. `AnalysisRun`s have a lifecycle (e.g. `Running`, `Successful`, `Failed`) and also record measurements taken during the run. `AnalysisRun`s appear in the Kargo UI under the <Hlt>Verifications</Hlt> tab of a `Stage`:
```

  ownerReferences:
    - apiVersion: kargo.akuity.io/v1alpha1
      blockOwnerDeletion: false
      kind: Freight
      name: 319ddec6bbfbd808d499a0357b027e773aa3707f
      uid: fb0db653-92af-420b-9092-0c7bf9be290e
  resourceVersion: "54013326"
  uid: 8cb745cb-554b-4141-80ee-e0df181cd56e
spec:
  metrics:
  - name: integration-test
    provider:
      job:
        spec:
          template:
            spec:
              containers:
              - name: sleep
                image: alpine:latest
                command: [sleep, "10"]
              restartPolicy: Never
          backoffLimit: 1
status:
  dryRunSummary: {}
  message: Metric "integration-test" assessed Failed due to failed (1) > failureLimit (0)
  metricResults:
    - count: 1
      failed: 1
      measurements:
        - finishedAt: 2025-01-28T20:50:31Z
          metadata:
            job-name: 8cb745cb-554b-4141-80ee-e0df181cd56e.integration-test.1
            job-namespace: guestbook
          phase: Failed
          startedAt: 2025-01-28T20:47:50Z
      name: integration-test
      phase: Failed
  phase: Failed
  runSummary:
    count: 1
    failed: 1
  startedAt: 2025-01-28T20:47:50Z
```

:::note

> The difference between an `AnalysisTemplate` versus an `AnalysisRun` is that an `AnalysisTemplate` is the _definition_ how to perform verification, and the `AnalysisRun` tracks the _progress and result_ of an run of one or more template(s).

Reviewer: ```suggestion
__`AnalysisTemplate` vs `AnalysisRun`__

An `AnalysisTemplate` is _defines_ a verification process, while an `AnalysisRun` tracks the _progress and result_ of an execution of that process.
```

:::

##  Arguments and Metadata

It is also possible to specify additional labels, annotations, and arguments
that should be applied to `AnalysisRun` resources spawned from the referenced
`AnalysisTemplate`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  # ...
  verification:
    analysisTemplates:
    - name: kargo-demo
    analysisRunMetadata:
      labels:
        foo: bar
      annotations:
        bat: baz
    args:
    - name: foo
      value: bar
```

## Implicit Argo CD Verification

> It is usually the desire to wait until a deployment is fully rolled out and updated, before performing any testing and verification. For example, it would be premature to start running tests if a Deployment's Pods were still coming up and or not fully updated to the desired version. Promotions might only include the apply/sync of new manifests to the cluster, but do not not necessarily wait until the rollout was fully complete (which might take minutes depending on many factors like Pod startup time).

Reviewer: Note to self that @jessesuen and I have a DM on this.

> It is usually the desire to wait until a deployment is fully rolled out and updated, before performing any testing and verification. For example, it would be premature to start running tests if a Deployment's Pods were still coming up and or not fully updated to the desired version. Promotions might only include the apply/sync of new manifests to the cluster, but do not not necessarily wait until the rollout was fully complete (which might take minutes depending on many factors like Pod startup time).

Reviewer: ```suggestion
It is usually desirable for changes to be fully rolled out before performing any verification. For example, it would be premature to start running tests if a `Deployment`'s `Pod`s were still starting or not fully transitioned to the desired version. `Promotion`s that integrate with Argo CD do wait for any sync operations they initiate to complete, but it is possible for an Argo CD `Application` to still be `Progressing` even after sync operations have completed and the `Promotion` process is considered complete.
```

> In order to prevent verification from starting before Applications are ready, Kargo has a built in understanding of Argo CD health, and will delay beginning verification until the Application had reached a Healthy state.

Reviewer 2: ```suggestion
In order to prevent verification from starting before Applications are ready, Kargo has a built-in understanding of [Argo CD health](https://argo-cd.readthedocs.io/en/latest/operator-manual/health/), and will delay beginning verification until the Application had reached a Healthy state.
```

> In order to prevent verification from starting before Applications are ready, Kargo has a built in understanding of Argo CD health, and will delay beginning verification until the Application had reached a Healthy state.

Reviewer: ```suggestion
In order to prevent verification from starting before `Application`s are ready, Kargo has a built-in understanding of Argo CD `Application` health and will delay verification until `Application`s have reached a `Healthy` state.
```

> If a Stage references one ore more Argo CD Applications as part of its promotion steps, after Promotion, Kargo will automatically wait until the Argo CD Applications reach a `Healthy` state. This is an built-in, "implicit" form of verification that is not configured explicitly in the `spec.verifications`. After the Applications reaches the `Healthy` state, Kargo will then run the specified verifications in `spec.verifications` (if any). This ensures that testing begins only after all resources of the Application have been successfully updated after the promotion.

Reviewer: ```suggestion
:::note
If a `Stage` references one ore more Argo CD `Application`s as part of its promotion process, but does not explicitly define any verification process, successful verification is implicitly contingent on the `Application`s reaching a `Healthy` state.
:::
```

## Soak Times

> In some scenarios, automated testing may be unavailable or insufficient to ensure confidence before promoting software to critical environments. To mitigate this, users often resort to letting the release run in a lower environment for a period of time, monitoring its behavior and stability before proceeding with promotion. This practice is commonly known as 'soaking' or 'baking' the release before it reaches production.

Reviewer: ```suggestion
In some scenarios, automated testing may be unavailable or insufficient to ensure confidence before promoting software to critical environments. To mitigate this, users often resort to letting the release run in a lower environment for a period of time, monitoring its behavior and stability before proceeding with promotion. This practice is commonly known as "soaking" or "baking" the release before it reaches production.
```

> Kargo features a "soak time" option in Stages when it requests Freight from upstream stages. To configure a soak time, a duration string can `spec.requestedFreight[].requiredSoakTime`. Valid durations are in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (beginning after successful promotion), for which the requested Freight must have continuously occupied ("soaked in") in an upstream Stage before becoming available for promotion to this Stage. Automated and manual promotions would be blocked until the soakTime requirement was met.

Reviewer 2: ```suggestion
Stages in Kargo can enforce a minimum "soak time" for Freight requested from upstream stages. The soak time is specified in spec.requestedFreight[].requiredSoakTime using a duration string with values in seconds, minutes, or hours (e.g., 180s, 30m, 48h). The soak time begins after successful promotion and represents how long Freight must remain in an upstream Stage before it becomes eligible for promotion to the requesting Stage. Both automated and manual promotions are blocked until this requirement is met.
```

> Kargo features a "soak time" option in Stages when it requests Freight from upstream stages. To configure a soak time, a duration string can `spec.requestedFreight[].requiredSoakTime`. Valid durations are in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (beginning after successful promotion), for which the requested Freight must have continuously occupied ("soaked in") in an upstream Stage before becoming available for promotion to this Stage. Automated and manual promotions would be blocked until the soakTime requirement was met.

Reviewer 2: ```suggestion
Kargo features a "soak time" option in Stages when it requests Freight from upstream stages. To configure a soak time, a duration string can be set at `spec.requestedFreight[].requiredSoakTime`. Valid durations are in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (beginning after successful promotion), for which the requested Freight must have continuously occupied ("soaked in") in an upstream Stage before becoming available for promotion to this Stage. Automated and manual promotions would be blocked until the soak time requirement was met.
```

or maybe even:

```suggestion
Stages in Kargo can enforce a minimum "soak time" for Freight requested from upstream stages. The soak time is specified in spec.requestedFreight[].requiredSoakTime using a duration string with values in seconds, minutes, or hours (e.g., 180s, 30m, 48h). The soak time begins after successful promotion and represents how long Freight must remain in an upstream Stage before it becomes eligible for promotion to the requesting Stage. Both automated and manual promotions are blocked until this requirement is met.
```

Reviewer: This is another I had revised prior to noticing the comment. Happy to revise further if you think it necessary.

> Kargo features a "soak time" option in Stages when it requests Freight from upstream stages. To configure a soak time, a duration string can `spec.requestedFreight[].requiredSoakTime`. Valid durations are in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (beginning after successful promotion), for which the requested Freight must have continuously occupied ("soaked in") in an upstream Stage before becoming available for promotion to this Stage. Automated and manual promotions would be blocked until the soakTime requirement was met.

Reviewer 2: > [..] and manual promotions would be blocked [..]

Should we hint here that it can always be overwritten using manual _approval_? I think people otherwise may think there is no way out if you want to.

Reviewer: I did add that below this section.

> Kargo features a "soak time" option in Stages when it requests Freight from upstream stages. To configure a soak time, a duration string can `spec.requestedFreight[].requiredSoakTime`. Valid durations are in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (beginning after successful promotion), for which the requested Freight must have continuously occupied ("soaked in") in an upstream Stage before becoming available for promotion to this Stage. Automated and manual promotions would be blocked until the soakTime requirement was met.

Reviewer: ```suggestion
`Stage`s may optionally specify a "soak time" when requesting `Freight` from upstream `Stage`s. To configure this, a duration can be specified using the `spec.requestedFreight[].requiredSoakTime` field. Valid durations are expressed in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (following a successful promotion), for which the requested `Freight` must have continuously occupied ("soaked in") in an upstream `Stage` before becoming available for promotion to this `Stage`. Both automated and manual promotions are blocked until the soak time requirement is met.
```

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: prod
  namespace: guestbook
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: guestbook
    sources:
      stages:
      - uat
    requiredSoakTime: 1h
 # Omitted for brevity...
```

:::note

> Soak Time is not technically a verification feature, but is documented alongside verification since both features are closely related and may be used in conjunction with each other. Both relate to prevention of Freight promotion from upstream Stages based on required conditions and behavior.

Reviewer 2: I think the following bears the same message while being a bit shorter/easier to read:

```suggestion
While soak time is distinct from verification, these features are documented together as they serve complementary purposes in preventing premature Freight promotion between upstream Stages based on specific conditions.
```

Reviewer: I had already revised the wording to something similar to your suggestion before I noticed this. Let me know if you think it needs further improvement.

Reviewer: Alternatively, please feel free to edit it directly. @jessesuen said it was ok.

> Soak Time is not technically a verification feature, but is documented alongside verification since both features are closely related and may be used in conjunction with each other. Both relate to prevention of Freight promotion from upstream Stages based on required conditions and behavior.

Reviewer: ```suggestion
Soak Time is not technically a verification feature, but is documented alongside here because both relate to preventing `Freight` from moving downstream until required conditions have been met. These two features may even be used in conjunction with one another.
```

:::

> 

Reviewer: ```suggestion
:::info
Manually approving `Freight` for a `Stage` makes it immediately available to that `Stage` _regardless_ of whether any required soak time has elapsed.
:::
```

## ClusterAnalysisTemplates

> References to ClusterAnalysisTemplates are currently unsupported but is expected to be in a future release.

Reviewer: ```suggestion
Referencing to `ClusterAnalysisTemplate`s is currently unsupported but is expected to be in a future release.
```

---
description: Learn how to verify a Stage after Promotion

> sidebar_label: Verifying Freight in a Stage

Reviewer: ```suggestion
sidebar_label: Verifying Freight
```

---

> # Verifying a Stage after Promotion

Reviewer: ```suggestion
# Verifying Freight
```

> After successfully promoting `Freight` to a `Stage`, it is often desired to verify the update was successful. Verification is an optional, user-defined process that qualifies `Freight` in a `Stage`. Common examples of verification include:

Reviewer: ```suggestion
Following a successful `Promotion` (i.e. the _process_ succeeded), it is often desired to further verify that updated code or configuration meets expectations. Verification is an optional, user-defined process that asserts `Freight` in a `Stage`. Common examples of verification include:
```

> Following a successful `Promotion` (i.e. the _process_ succeeded), it is often desired to further verify that updated code or configuration meets expectations. Verification is an optional, user-defined process that asserts `Freight` in a `Stage`. Common examples of verification include:

Reviewer: ```suggestion
Following a successful `Promotion` (i.e. the _process_ succeeded), it is often desired to assert whether updated code or configuration meets expectations. Verification is an optional, user-defined process to test the `Freight` in a given `Stage`. Common examples of verification include:
```

* Executing containerized integration test scripts
* Querying monitoring tools (e.g. DataDog, Prometheus) and measuring application KPIs

> * Performing HTTP requests to internal services or systems and interpreting response as success/fail

Reviewer: ```suggestion
* Performing HTTP requests to internal services or systems and interpreting responses as success/fail
```

* Running load generation tests

When a `Stage` with verification is upstream from other `Stages`, `Freight` must pass verification there before becoming eligible to be promoted downstream.

## Configuring Verification

A `Stage` can be configured with verification by referencing one or more `AnalysisTemplates` in its `spec.verification` field. In the following example, the `dev` `Stage` references an `AnalysisTemplate` named `integration-test`, which executes after any successful `Promotion`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: dev
  namespace: guestbook
spec:
  # ...
  verification:
    analysisTemplates:
    - name: integration-test
```

## Configuring AnalysisTemplates

> An [AnalysisTemplate](../60-reference-docs/50-analysis-templates.md) defines instructions for how to perform verification. It might include running a Kubernetes `Job`, querying monitoring tools, interpreting their result, performing HTTP requests, or any combination of these actions. `AnalysisTemplate`s reside in the same `Project` namespace as the `Stage` from which they are referenced. The following simple `AnalysisTemplate` example will create a Kubernetes `Job` that runs an alpine container that sleeps for 10 seconds:

Reviewer: ```suggestion
An [AnalysisTemplate](../60-reference-docs/50-analysis-templates.md) defines instructions for how to perform verification. It might include running a Kubernetes `Job`, querying monitoring tools, interpreting their result, performing HTTP requests, or any combination of these actions. `AnalysisTemplate`s reside in the same project namespace as the `Stage` from which they are referenced. The following simple `AnalysisTemplate` example will create a Kubernetes `Job` that runs an alpine container that sleeps for 10 seconds:
```

```yaml
apiVersion: argoproj.io/v1alpha1
kind: AnalysisTemplate
metadata:
  name: integration-test
  namespace: guestbook
spec:
  metrics:
  - name: integration-test
    provider:
      job:
        spec:
          template:
            spec:
              containers:
              - name: sleep
                image: alpine:latest
                command: [sleep, "10"]
              restartPolicy: Never
          backoffLimit: 1
```

:::info
For further documentation and examples of `AnalysisTemplate`s, refer to the
[`AnalysisTemplate` reference](../60-reference-docs/50-analysis-templates.md).
:::

## How Verification Works

> Upon successful promotion, a Stage enters the `Verifying` phase. Any AnalysisTemplates which were referenced in the Stage are invoked into an resource called an `AnalysisRun`, which represents a single invocation of the templates into a batch-like object. 

Reviewer: ```suggestion
After a successful `Promotion`, a `Stage` enters the `Verifying` phase. Any `AnalysisTemplate`s which were referenced by the `Stage` are are used to spawn a resource called an `AnalysisRun`, which represents a single execution of those templates. 
```

> When the `AnalysisRun` completes and is `Successful`, the verification is considered successful and the `Freight` is marked as "verified" in that `Stage`, which makes it available for promotion to any immediate downstream `Stage`s, either automatically (with auto-promotion) or manually (via UI, API, CLI). `Freight` which have not passed verification in a `Stage` are blocked continuing further downstream.

Reviewer: ```suggestion
When the `AnalysisRun` completes and is `Successful`, the verification is considered successful and the `Freight` is marked as "verified" in that `Stage`, which makes it available for promotion to any immediate downstream `Stage`s, either automatically (with auto-promotion) or manually (via UI, API, CLI). `Freight` which have not passed verification in a `Stage` are blocked from continuing further downstream.
```

:::tip
It is sometimes desired to skip the verification process entirely (e.g. a hotfix needs to be fast-tracked to production). Verification (and the entire Pipeline for that matter) can be bypassed by [manually approving](./15-working-with-freight.md#manual-approvals) `Freight` for a specified `Stage`.
:::

While a `Stage` is verifying (i.e. the `AnalysisRun` is not yet completed) no other `Promotion`s to that `Stage` will run until verification has completed (either successfully or unsuccessfully). Once verification has completed, the next queued `Promotion` will run.

## AnalysisRun

An `AnalysisRun` is a resource representing a verification attempt for a `Freight` + `Stage` pair. `AnalysisRun`s have a lifecycle (e.g. `Running`, `Successful`, `Failed`) and also record measurements taken during the run. `AnalysisRun`s appear in the Kargo UI under the <Hlt>Verifications</Hlt> tab of a `Stage`:

![AnalysisRun in Kargo UI](./img/analysisrun.png)

> The following might be an example of a `AnalysisRun` spawned from the `integration-test` template:

Reviewer: ```suggestion
The following might be an example of an `AnalysisRun` spawned from the `integration-test` template:
```

:::

##  Arguments and Metadata

It is also possible to specify additional labels, annotations, and arguments
that should be applied to `AnalysisRun` resources spawned from the referenced
`AnalysisTemplate`:

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: test
  namespace: kargo-demo
spec:
  # ...
  verification:
    analysisTemplates:
    - name: kargo-demo
    analysisRunMetadata:
      labels:
        foo: bar
      annotations:
        bat: baz
    args:
    - name: foo
      value: bar
```

## Implicit Argo CD Verification

It is usually desirable for changes to be fully rolled out before performing any verification. For example, it would be premature to start running tests if a `Deployment`'s `Pod`s were still starting or not fully transitioned to the desired version. `Promotion`s that integrate with Argo CD do wait for any sync operations they initiate to complete, but it is possible for an Argo CD `Application` to still be `Progressing` even after sync operations have completed and the `Promotion` process is considered complete.

> In order to prevent verification from starting before `Application`s are ready, Kargo has a built-in understanding of Argo CD `Application` health and will delay verification until `Application`s have reached a `Healthy` state.

Reviewer: ```suggestion
In order to prevent verification from starting before `Application`s are ready, Kargo has a built-in understanding of [Argo CD `Application` health](https://argo-cd.readthedocs.io/en/latest/operator-manual/health/) and will delay verification until `Application`s have reached a `Healthy` state.
```

`Stage`s may optionally specify a "soak time" when requesting `Freight` from upstream `Stage`s. To configure this, a duration can be specified using the `spec.requestedFreight[].requiredSoakTime` field. Valid durations are expressed in seconds, minutes or hours (e.g. `180s`, `30m`, `48h`). This duration is the minimum duration (following a successful promotion), for which the requested `Freight` must have continuously occupied ("soaked in") in an upstream `Stage` before becoming available for promotion to this `Stage`. Both automated and manual promotions are blocked until the soak time requirement is met.

```yaml
apiVersion: kargo.akuity.io/v1alpha1
kind: Stage
metadata:
  name: prod
  namespace: guestbook
spec:
  requestedFreight:
  - origin:
      kind: Warehouse
      name: guestbook
    sources:
      stages:
      - uat
    requiredSoakTime: 1h
 # Omitted for brevity...
```

:::note

> Soak Time is not technically a verification feature, but is documented alongside it because both relate to preventing `Freight` from moving downstream until required conditions have been met. These two features may even be used in conjunction with one another.

Reviewer: ```suggestion
Soak time is not technically a verification feature, but is documented alongside it because both relate to preventing `Freight` from moving downstream until required conditions have been met. These two features may even be used in conjunction with one another.
```

:::

:::info
[Manually approving](./15-working-with-freight.md#manual-approvals) `Freight` for a `Stage` makes it immediately available to that `Stage` _regardless_ of whether any required soak time has elapsed.
:::
## ClusterAnalysisTemplates

> Referencing `ClusterAnalysisTemplate`s is currently unsupported but is expected to be in a future release.

Reviewer: ```suggestion
Referencing `ClusterAnalysisTemplate`s is currently unsupported but is expected to be possible in a future release.
```
